cranelift/codegen/src/egraph/elaborate.rs

```diff
@@ -0,0 +1,589 @@
+        loop_analysis: &'a LoopAnalysis,
+        egraph: &'a EGraph<NodeCtx, Analysis>,
+        node_ctx: &'a NodeCtx,
+        remat_ids: &'a FxHashSet<Id>,
+        stats: &'a mut Stats,
+    ) -> Self {
+        let num_blocks = func.dfg.num_blocks();
+        let mut id_to_best_cost_and_node =
+            SecondaryMap::with_default((Cost::infinity(), Id::invalid()));
+        id_to_best_cost_and_node.resize(egraph.classes.len());
+        Self {
+            func,
+            domtree,
+            loop_analysis,
+            egraph,
+            node_ctx,
+            id_to_value: ScopedHashMap::with_capacity(egraph.classes.len()),
+            id_to_best_cost_and_node,
+            loop_stack: smallvec![],
+            cur_block: None,
+            first_branch: SecondaryMap::with_capacity(num_blocks),
+            remat_ids,
+            elab_stack: vec![],
+            elab_result_stack: vec![],
+            block_stack: vec![],
+            stats,
+        }
+    }
+
+    fn cur_loop_depth(&self) -> LoopDepth {
+        self.loop_stack.len() as LoopDepth
+    }
+
+    fn start_block(&mut self, idom: Option<Block>, block: Block, block_params: &[(Id, Type)]) {
+        trace!(
+            "start_block: block {:?} with idom {:?} at loop depth {} scope depth {}",
+            block,
+            idom,
+            self.cur_loop_depth(),
+            self.id_to_value.depth()
+        );
+
+        // Note that if the *entry* block is a loop header, we will
+        // not make note of the loop here because it will not have an
+        // immediate dominator. We must disallow this case because we
+        // will skip adding the `LoopStackEntry` here but our
+        // `LoopAnalysis` will otherwise still make note of this loop
+        // and loop depths will not match.
+        if let Some(idom) = idom {
+            if self.loop_analysis.is_loop_header(block).is_some() {
+                self.loop_stack.push(LoopStackEntry {
+                    // Any code hoisted out of this loop will have code
+                    // placed in `idom`, and will have def mappings
+                    // inserted in to the scoped hashmap at that block's
+                    // level.
+                    hoist_block: idom,
+                    scope_depth: (self.id_to_value.depth() - 1) as u32,
+                });
+                trace!(
+                    " -> loop header, pushing; depth now {}",
+                    self.loop_stack.len()
+                );
+            }
+        } else {
+            debug_assert!(
+                self.loop_analysis.is_loop_header(block).is_none(),
+                "Entry block (domtree root) cannot be a loop header!"
+            );
+        }
+
+        self.cur_block = Some(block);
+        for &(id, ty) in block_params {
+            let value = self.func.dfg.append_block_param(block, ty);
+            trace!(" -> block param id {:?} value {:?}", id, value);
+            self.id_to_value.insert_if_absent(
```

Comment:
Out of curiosity: since block params are unique, each such enode _should_ be absent at this point, right?

Reply:
Yep, but the only API we have is `insert_if_absent`, not `insert_if_absent_and_assert_otherwise` :-)